From 43364cbf9bb4ef97a4daf238c11ff52bbab2d7ac Mon Sep 17 00:00:00 2001
From: Xiaofei Tan <tanxiaofei@huawei.com>
Date: Mon, 23 Mar 2026 20:09:05 +0800
Subject: [PATCH 913/960] downstream: ACPI: APEI: Add vendor specific error
 info handling for ARM Processor

In the vendor specific error info of ARM Processor Error Section, Hisi
defines a critical flag. When a CPU core consumes memory fault and a
synchronization exception occurs. Set this flag if the memory fault is
device-level fault, such as a remote memory link disconnection, where
the CPU core may generate a large number of synchronization exceptions.
OS can handle the error more efficient with this flag.

Signed-off-by: Xiaofei Tan <tanxiaofei@huawei.com>
Signed-off-by: slim6882 <yangjunshuo@huawei.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 drivers/acpi/apei/Kconfig            |  9 +++
 drivers/acpi/apei/Makefile           |  1 +
 drivers/acpi/apei/apei-internal.h    | 11 ++++
 drivers/acpi/apei/ghes-vendor-info.c | 90 ++++++++++++++++++++++++++++
 4 files changed, 111 insertions(+)
 create mode 100644 drivers/acpi/apei/ghes-vendor-info.c

diff --git a/drivers/acpi/apei/Kconfig b/drivers/acpi/apei/Kconfig
index 71e0d64a7792..bc61f8ceb51f 100644
--- a/drivers/acpi/apei/Kconfig
+++ b/drivers/acpi/apei/Kconfig
@@ -34,6 +34,15 @@ config ACPI_APEI_GHES
 	  by firmware to produce more valuable hardware error
 	  information for Linux.
 
+config ACPI_APEI_GHES_ARMP_VENDOR_INFO
+	bool "ARM processor vendor specific error info"
+	depends on ARM64 && ACPI_APEI_GHES
+	default n
+	help
+	  Parse vendor specific error info of ARM processor error section,
+	  for platform-related fault status to handle system failures. This
+	  functionality requires support from both the chip and the firmware.
+
 config ACPI_APEI_PCIEAER
 	bool "APEI PCIe AER logging/recovering support"
 	depends on ACPI_APEI && PCIEAER
diff --git a/drivers/acpi/apei/Makefile b/drivers/acpi/apei/Makefile
index 4dfac2128737..dde1dba823d7 100644
--- a/drivers/acpi/apei/Makefile
+++ b/drivers/acpi/apei/Makefile
@@ -1,6 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0
 obj-$(CONFIG_ACPI_APEI)		+= apei.o
 obj-$(CONFIG_ACPI_APEI_GHES)	+= ghes.o
+obj-$(CONFIG_ACPI_APEI_GHES_ARMP_VENDOR_INFO)	+= ghes-vendor-info.o
 obj-$(CONFIG_ACPI_APEI_EINJ)	+= einj.o
 obj-$(CONFIG_ACPI_APEI_ERST_DEBUG) += erst-dbg.o
 
diff --git a/drivers/acpi/apei/apei-internal.h b/drivers/acpi/apei/apei-internal.h
index 67c2c3b959e1..1268cf9cbb05 100644
--- a/drivers/acpi/apei/apei-internal.h
+++ b/drivers/acpi/apei/apei-internal.h
@@ -10,6 +10,7 @@
 #include <linux/acpi.h>
 
 struct apei_exec_context;
+struct cper_sec_proc_arm;
 
 typedef int (*apei_exec_ins_func_t)(struct apei_exec_context *ctx,
 				    struct acpi_whea_header *entry);
@@ -130,4 +131,14 @@ static inline u32 cper_estatus_len(struct acpi_hest_generic_status *estatus)
 }
 
 int apei_osc_setup(void);
+
+#ifdef CONFIG_ACPI_APEI_GHES_ARMP_VENDOR_INFO
+bool ghes_armp_vendor_critical_error(struct cper_sec_proc_arm *err, bool sync);
+#else
+static inline bool
+ghes_armp_vendor_critical_error(struct cper_sec_proc_arm *err, bool sync)
+{
+	return false;
+}
+#endif
 #endif
diff --git a/drivers/acpi/apei/ghes-vendor-info.c b/drivers/acpi/apei/ghes-vendor-info.c
new file mode 100644
index 000000000000..a7878dfa2ac0
--- /dev/null
+++ b/drivers/acpi/apei/ghes-vendor-info.c
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Handle ARM processor vendor specific error info.
+ * Copyright (c) Huawei Technologies Co., Ltd. 2026. All rights reserved.
+ */
+
+#include <linux/init.h>
+#include <linux/acpi.h>
+#include <acpi/ghes.h>
+#include <acpi/apei.h>
+#include "apei-internal.h"
+
+#define HISI_OEM	BIT(0)
+
+static int vendor_oem __ro_after_init;
+
+#ifdef CONFIG_ARCH_HISI
+
+#define HISI_VENDOR_MAGIC_NUM		0xCC08CC08CC08CC08
+#define HISI_VENDOR_CRITICAL_ERR	BIT(0)
+
+struct hisi_armp_vendor_info {
+	u64 magic_num;
+	u32 ver_info;
+	u32 err_flag;	/* bit0:critical error, others: reserved */
+	u32 *regs;
+} __packed;
+
+static bool ghes_hisi_critical_hw_error(struct cper_sec_proc_arm *err, bool sync)
+{
+	struct hisi_armp_vendor_info *vendor_info;
+	unsigned long err_info_sz;
+	char *p;
+
+	if (!sync)
+		return false;
+
+	if (!(err->validation_bits & CPER_ARM_VALID_VENDOR_INFO))
+		return false;
+
+	p = (char *)(err + 1);
+	err_info_sz = sizeof(struct cper_arm_err_info) * err->err_info_num;
+	if (!err->context_info_num) {
+		vendor_info = (struct hisi_armp_vendor_info *)
+			(p + err_info_sz);
+	} else {
+		struct cper_arm_ctx_info *ctx_info = (struct cper_arm_ctx_info *)
+			(p + err_info_sz);
+
+		vendor_info = (struct hisi_armp_vendor_info *)
+			(p + err_info_sz +
+			ctx_info->size * err->context_info_num);
+	}
+
+	if (vendor_info->magic_num != HISI_VENDOR_MAGIC_NUM)
+		return false;
+
+	return (bool)(vendor_info->err_flag & HISI_VENDOR_CRITICAL_ERR);
+}
+#else
+static inline bool ghes_hisi_critical_hw_error(struct cper_sec_proc_arm *err, bool sync)
+{
+	return false;
+}
+#endif
+
+bool ghes_armp_vendor_critical_error(struct cper_sec_proc_arm *err, bool sync)
+{
+	if (vendor_oem & HISI_OEM)
+		return ghes_hisi_critical_hw_error(err, sync);
+
+	return false;
+}
+
+static int __init ghes_check_oem_table(void)
+{
+	struct acpi_table_header *tbl;
+	acpi_status status = AE_OK;
+
+	status = acpi_get_table(ACPI_SIG_HEST, 0, &tbl);
+	if (ACPI_FAILURE(status) || !tbl)
+		return -ENODEV;
+
+	if (!memcmp(tbl->oem_id, "HISI  ", ACPI_OEM_ID_SIZE))
+		vendor_oem |= HISI_OEM;
+
+	acpi_put_table(tbl);
+	return 0;
+}
+subsys_initcall(ghes_check_oem_table);
-- 
2.43.0

